pgjdbc/src/main/java/com/yugabyte/ysql/YBjdbcHostnameVerifier.java

```diff
@@ -0,0 +1,225 @@
+
+  private static final Logger LOGGER = Logger.getLogger(YBjdbcHostnameVerifier.class.getName());
+  protected static final String GET_SERVERS_QUERY = "select * from yb_servers()";
+  protected Boolean useHostColumn = null;
+
+
+
+
+  private static final int TYPE_DNS_NAME = 2;
+  private static final int TYPE_IP_ADDRESS = 7;
+  private final Properties originalProperties;
+  protected ArrayList<String> currentPublicIps = new ArrayList<>();
+  protected Map<String, String> hostPortMap = new HashMap<>();
+  protected Map<String, String> hostPortMapPublic = new HashMap<>();
+  private final PGStream stream;
+
+  public YBjdbcHostnameVerifier(Properties props, PGStream stream){
+    this.originalProperties = props;
+    this.stream= stream;
+
+  }
+
+  @Override
+  public boolean verify(String hostname, SSLSession session) {
+
+    X509Certificate[] peerCerts;
+    try {
+      peerCerts = (X509Certificate[]) session.getPeerCertificates();
+    } catch (SSLPeerUnverifiedException e) {
+      LOGGER.log(Level.SEVERE,
+          GT.tr("Unable to parse X509Certificate for hostname {0}", hostname), e);
+      return false;
+    }
+    if (peerCerts == null || peerCerts.length == 0) {
+      LOGGER.log(Level.SEVERE,
+          GT.tr("No certificates found for hostname {0}", hostname));
+      return false;
+    }
+
+    X509Certificate serverCert = peerCerts[0];
+
+    // Check for Subject Alternative Names (see RFC 6125)
+
+    Collection<List<?>> subjectAltNames;
+    try {
+      subjectAltNames = serverCert.getSubjectAlternativeNames();
+      if (subjectAltNames == null) {
+        subjectAltNames = Collections.emptyList();
+      }
+    } catch (CertificateParsingException e) {
+      LOGGER.log(Level.SEVERE,
+          GT.tr("Unable to parse certificates for hostname {0}", hostname), e);
+      return false;
+    }
+
+    boolean anyDnsSan = false;
+    String san = null;
+    /*
+     * Each item in the SAN collection is a 2-element list.
+     * See {@link X509Certificate#getSubjectAlternativeNames}
+     * The first element in each list is a number indicating the type of entry.
+     */
+    for (List<?> sanItem : subjectAltNames) {
+      if (sanItem.size() != 2) {
+        continue;
+      }
+      Integer sanType = (Integer) sanItem.get(0);
+      if (sanType == null) {
+        // just in case
+        continue;
+      }
+      if (sanType != TYPE_IP_ADDRESS && sanType != TYPE_DNS_NAME) {
+        continue;
+      }
+      san = (String) sanItem.get(1);
+      if (sanType == TYPE_IP_ADDRESS && san != null && san.startsWith("*")) {
+        // Wildcards should not be present in the IP Address field
+        continue;
+      }
+      anyDnsSan |= sanType == TYPE_DNS_NAME;
+    }
+
+    Connection controlConnection = null;
+    originalProperties.setProperty("PGHOST", san);
+    HostSpec[] hspec = hostSpecs(this.originalProperties);
+    try {
+      controlConnection = new PgConnection(
```

Comment:
We should keep the connection around for some time instead of creating it for each connection verification.